Support integer variables

mtanneau opened the pull request:
Addresses #28.

Summary of changes:
* `QPSData` now has a `vartypes` field that contains the type of each variable
* Integer markers are checked

dpo commented on src/readqps.jl:
```diff
@@ -124,6 +127,14 @@ function row_type(rtype::String)
+const VTYPE_B =  1  # Binary variable
+const VTYPE_I =  2  # Integer variable
+const VTYPE_S =  3  # Semi-continuous
+const VTYPE_N =  4  # Semi-integer
```
Are these values in the MPS "standard" somehow or did you make them up? Perhaps using symbols would be somewhat more explicit.

mtanneau replied:
The integer values themselves are made-up.
The significance matches Gurobi's [variable types](https://www.gurobi.com/documentation/9.0/refman/vtype.html).

I added the "marked integer" category to make the parsing of bounds easier (see my comments in #28 regarding how solvers have different conventions even for that).

mtanneau replied:
Symbols, or `Char`s would indeed make it easier for people to work with.

dpo replied:
Maybe use a Dict{Symbol,Int} that maps human-readable symbols to the Gurobi ints?

odow replied:
A better option would be to make this an `Enum`?
```Julia
@enum(
    VariableType,
   VTYPE_ MARKED_INTEGER,
   VTYPE_ CONTINUOUS,
   VTYPE_ BINARY,
  ...
)
```

mtanneau replied:
The issue I have with `Enum`s is that users then have to work with the EnumType, and most likely have to translate it.
I think `Char`s are more flexible and easier to work with. Also, it doesn't hurt that a number of solvers follow a similar convention ('C' for continuous, 'B' for binary, etc...).

odow replied:
`Cchar`? Following Gurobi/CPLEX convention is probably okay.

dpo commented:
I'm not at all familiar with the way integer variables are represented in MPS, so I'll trust you.

odow commented on src/readqps.jl:
```diff
@@ -800,7 +808,18 @@ function readqps(qps::IO; mpsformat::Symbol=:free)
+                elseif card.f3 == "'INTEND'"
+                    integer_section = false
+                else
+                    @error "Ignoring marker $(card.f3) at line $(card.nline)"
```
Shouldn't this just be `@warn`?

mtanneau replied:
The convention we adopted is that rim data trigger an `@error` log message.
Such messages can be suppressed by setting the logger as desired.

odow replied:
Ohhhh. I was wondering why I didn't see the errors when reading SMIPs. I forgot about the `NullLogger` in SMPSReader.

odow commented on src/readqps.jl:
```diff
@@ -816,5 +835,32 @@ function readqps(qps::IO; mpsformat::Symbol=:free)
 
+    # Finalize variable bounds
+    # All marked integer variables with no explicit bounds
+    # are converted to Integer variables with bounds [0, 1]
```
I think this is solver-dependent. Options being `(-Inf, Inf)`, `[0, Inf)`, or `[0, 1]`. Seems like it should probably be `[0, Inf)`, but we should throw a warning. This is messy.

mtanneau commented:
I fixed a small issue in finalizing bounds, and I added more tests.

The convention should be as follows:

| Marker? | `BOUNDS` fields | Type | Bounds reported |
|:--:|:--:|:--:|:--:|
| Yes | - | Integer | `[0, 1]` 
| Yes | `BV` | Binary | `[0, 1]` 
| Yes | (`LI`, `l`) | Integer | `[l, Inf]` 
| Yes | (`UI`, `u`) with `u≥0` | Integer | `[0, u]`
| Yes | (`UI`, `u`) with `u<0` | Integer | `[-Inf, u]`
| Yes | (`LI`, `l`) + (`UI`, `u`) | Integer | `[l, u]` 
| No | `BV` | Binary | `[0, 1]` 
| No | (`LI`, `l`) | Integer | `[l, Inf]` 
| No | (`UI`, `u`) with `u≥0` | Integer | `[0, u]`
| No | (`UI`, `u`) with `u<0` | Integer | `[-Inf, u]`
| No | (`LI`, `l`) + (`UI`, `u`) | Integer | `[l, u]` 

The `LI`/`UI` can be replaced by `LO`/`UP` in the table above, with no impact on bounds. Only the integrality of variables are affected.

mtanneau commented:
That should do it.

README is up-to-date.

Since we're using `Char`s for variable types, I also switched to `Char`s for constraint senses.
This is a breaking change.

mtanneau commented:
> Since we're using `Char`s for variable types, I also switched to `Char`s for constraint senses.
> This is a breaking change.
@dpo any objection to this?

If not, I'll merge this tomorrow, and tag v0.2.0

dpo commented:
Personally, I'm in favor of symbols rather than chars as they represent a unique object.

mtanneau commented:
Last call for objections.
@odow I changed from `Char`s to `Symbol`s; the correspondence is in the README.

odow commented:
I think we should either use `Cchar` so they can be given to CPLEX/Gurobi without copying, or `Enum` so it isn't possible to have typos, etc. 

But I don't care enough to block the PR.

mtanneau commented:
Given that Mosek and GLPK do not use `Cchar`s (it looks like it's a CPLEX/Gurobi convention), I decided to go with `Enum`s.
As @odow pointed out, it's less prone to typos and other mistakes, and it should make it easier to track down such bugs.
